File: app/src/main/java/com/onedive/passwordstore/presentation/activity/BaseSecurityActivity.java
package com.onedive.passwordstore.presentation.activity;

import static com.onedive.passwordstore.utils.SharedPreferenceKt.getDefaultSharedPreference;

import android.app.KeyguardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.viewbinding.ViewBinding;

import com.onedive.passwordstore.R;
import com.onedive.passwordstore.utils.Const;

/**
 * This class is used if in settings,
 * the application lock feature is activated and the device has an active biometric password/password,
 * then this class will display a dialog that is used to confirm this device password.
 *
 * @param <B> viewBinding to use
 * @see BaseActivity
 */
public abstract class BaseSecurityActivity<B extends ViewBinding> extends BaseActivity<B> {

    private KeyguardManager keyguardManager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        keyguardManager = (KeyguardManager) getSystemService(Context.KEYGUARD_SERVICE);
        lockAppAndDetailNavigation();
    }

    /**
     * Check whether the preference with the key Const.LOCK_APP_KEY_PREFERENCE exists / does not exist
     * if there is then run the startResultLauncher() method otherwise run the deviceSecurityIsNotAvailable() method
     *
     * @see Const
     */
    private void lockAppAndDetailNavigation() {

        if (getDefaultSharedPreference(this).getBoolean(Const.LOCK_APP_KEY_PREFERENCE,false)) {
            startResultLauncher();
        }else {
             deviceSecurityIsNotAvailable();
        }

    }

    /**
     * This method is used to check whether the device has an active password / biometric
     * If there is, then display the password confirmation dialog
     *
     * @see KeyguardManager
     */
    private void startResultLauncher() {
        if (keyguardManager.isKeyguardSecure()){
            Intent intent = keyguardManager.createConfirmDeviceCredentialIntent(getString(R.string.enter_device_password), getString(R.string.enter_device_password_desc));
            activityResultLauncher.launch(intent);
        }else{
            deviceSecurityIsNotAvailable();
        }
    }


    /**
     * if the result of the password entered is wrong then close this activity if not,
     * then run the deviceSecurityIsAvailable() method
     */
    private final ActivityResultLauncher<Intent> activityResultLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() != BaseActivity.RESULT_OK) {
                    finish();
                } else {
                    deviceSecurityIsAvailable();
                }
            }
    );

    /**
     * abstract method that is executed if the device has a password
     */
    protected abstract void deviceSecurityIsAvailable();

    /**
     * abstract method that is executed if the device does not have a password
     */
    protected abstract void deviceSecurityIsNotAvailable();
}

File: app/src/main/java/com/onedive/passwordstore/presentation/activity/BaseSecurityActivity.kt
package com.onedive.passwordstore.presentation.activity

import android.app.KeyguardManager
import android.content.Context
import android.os.Build
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
import androidx.biometric.BiometricPrompt
import androidx.biometric.BiometricPrompt.ERROR_NO_BIOMETRICS
import androidx.biometric.BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL
import androidx.biometric.BiometricPrompt.PromptInfo
import androidx.viewbinding.ViewBinding
import com.onedive.passwordstore.R
import com.onedive.passwordstore.utils.showToast

/**
 * This class is an abstract class that is used to request a password/biometric when an activity is
 * running
 * @see BaseActivity
 * @see KeyguardManager
 * @see BiometricPrompt
 * @see PromptInfo
 * @param B View Binding to use
 */
abstract class BaseSecurityActivity<B : ViewBinding> : BaseActivity<B>() {

    private lateinit var biometricPrompt: BiometricPrompt
    private lateinit var biometricPromptBuilder: PromptInfo


    @Suppress("DEPRECATION")
    protected fun showConfirmDialogWithAvailablePasswordOrBiometricPassword() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {

            biometricPromptBuilder = PromptInfo.Builder().apply {
                setTitle(getString(R.string.app_name))
                setDescription(getString(R.string.enter_device_password_desc))
                setConfirmationRequired(false)
                setAllowedAuthenticators(BIOMETRIC_STRONG or DEVICE_CREDENTIAL)
            }.build()

            biometricPrompt = BiometricPrompt(this, authCallback)
            biometricPrompt.authenticate(biometricPromptBuilder)

        } else {

            val keyguardManager = getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
            val intent = keyguardManager.createConfirmDeviceCredentialIntent(
                getString(R.string.app_name),
                getString(R.string.enter_device_password_desc)
            )

            val resultLauncher = registerForActivityResult(ActivityResultContracts
                .StartActivityForResult()) {

                if (it.resultCode == RESULT_OK) {
                    confirmPasswordOrBiometricPasswordIsSuccessfully()
                } else {
                    // If the back button is pressed or another action occurs that "Keyguard"
                    // fails to execute, complete the activity
                    finish()
                }

            }
            resultLauncher.launch(intent)

        }

    }

    // Biometric callback
    private val authCallback = object : BiometricPrompt.AuthenticationCallback() {


        override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
            /*
             If the error is not because the device does not have an active password / biometric,
              then display a toast, de authenticate, and close the activity
            */
            if (errorCode == ERROR_NO_BIOMETRICS || errorCode == ERROR_NO_DEVICE_CREDENTIAL) {
                confirmPasswordOrBiometricPasswordIsSuccessfully()
            } else {
                biometricPrompt.cancelAuthentication()
                showToast(
                    context = applicationContext,
                    message = "error with code $errorCode${
                        if (errString.isNotEmpty()) " with message $errString"
                        else ""
                    }",
                    duration = Toast.LENGTH_SHORT
                )
                finish()
            }
        }

        override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
            /*
             If authentication is successful,
             then execute the abstract function confirmPasswordOrBiometricPasswordIsSuccessfully()
             */
            confirmPasswordOrBiometricPasswordIsSuccessfully()
        }

        override fun onAuthenticationFailed() {
            // Display a toast with a message that you are not the original owner of this device
            showToast(
                context = applicationContext,
                message = "It looks like you are not the original owner of this device",
                duration = Toast.LENGTH_SHORT
            )
        }

    }


    /**
     * This function is performed when the user's confirmation with password / biometric
     * has been successful and if this device not available password
     */
    abstract fun confirmPasswordOrBiometricPasswordIsSuccessfully()

    /**
     * This function is performed when checking has been completed but the device
     * does not have an active password / biometric
     */
    abstract fun noAvailablePasswordOrBiometricPasswordInThisDevice()

}

